File: llmsim/sim/rng.py
"""Named, independently seeded random streams.

Why a registry instead of one global generator: if arrivals, prompt lengths and
failure injection all draw from one stream, then changing the arrival rate
changes every other draw too. Comparing two policies would then be comparing two
different workloads, and any difference you measure is partly noise.

With named streams, ``rng.stream("arrival")`` is unaffected by how many draws
``rng.stream("prompt_len")`` has taken. Turning on failure injection does not
perturb the workload. That is what makes A/B comparison in a simulator valid.

Stream seeds derive from ``blake2b(name, key=master_seed)``, not from Python's
``hash()``, which is randomized per process for strings and would destroy
reproducibility across runs.
"""

from __future__ import annotations

import hashlib
import math
import random

_PERSON = b"lbsim.rng"


class RngRegistry:
    """Creates and caches one :class:`random.Random` per name."""

    def __init__(self, master_seed: int) -> None:
        self.master_seed = int(master_seed)
        self._key = self.master_seed.to_bytes(8, "little", signed=False)
        self._streams: dict[str, random.Random] = {}

    def derive_seed(self, name: str) -> int:
        digest = hashlib.blake2b(
            name.encode("utf-8"), digest_size=8, key=self._key, person=_PERSON
        ).digest()
        return int.from_bytes(digest, "little")

    def stream(self, name: str) -> random.Random:
        rng = self._streams.get(name)
        if rng is None:
            rng = random.Random(self.derive_seed(name))
            self._streams[name] = rng
        return rng

    def child(self, name: str) -> "RngRegistry":
        """A sub-registry, so per-replica or per-tenant streams stay independent."""
        return RngRegistry(self.derive_seed(name))

    def __repr__(self) -> str:
        return f"RngRegistry(master_seed={self.master_seed})"


# ---------------------------------------------------------------------------
# Distributions
#
# Parameterised the way a workload is actually described: by a mean and a
# spread you can read off a dashboard, not by the shape parameters of the
# underlying math.
# ---------------------------------------------------------------------------


def exponential(rng: random.Random, mean: float) -> float:
    """Interarrival time of a Poisson process with rate ``1/mean``."""
    if mean <= 0:
        return 0.0
    return rng.expovariate(1.0 / mean)


def lognormal(rng: random.Random, mean: float, cv: float) -> float:
    """Lognormal with the given arithmetic ``mean`` and coefficient of variation.

    ``cv`` is stddev/mean. Token counts and service times are strongly
    right-skewed in practice, so lognormal is the sane default; ``cv`` around
    1.0 to 2.0 matches published prompt and completion length distributions
    better than an exponential does.
    """
    if mean <= 0:
        return 0.0
    if cv <= 0:
        return mean
    sigma2 = math.log1p(cv * cv)
    sigma = math.sqrt(sigma2)
    mu = math.log(mean) - sigma2 / 2.0
    return math.exp(rng.normalvariate(mu, sigma))


def pareto(rng: random.Random, scale: float, alpha: float) -> float:
    """Heavy tail. ``alpha`` <= 2 gives infinite variance, <= 1 infinite mean.

    Use for the long-context tail: a small fraction of requests with enormous
    prompts is exactly the load pattern that breaks KV-cache capacity planning.
    """
    return scale * (rng.paretovariate(alpha))


def bounded_int(value: float, lo: int, hi: int) -> int:
    """Clamp and round a continuous draw into a token count."""
    return max(lo, min(hi, int(round(value))))


def zipf(rng: random.Random, n: int, s: float = 1.0) -> int:
    """Draw in ``[0, n)`` with Zipf-like popularity, index 0 most popular.

    Models skewed key popularity: which system prompt, which tenant, which
    cached prefix. Real traffic is very skewed and uniform choice will
    understate prefix-cache hit rates badly.
    """
    if n <= 1:
        return 0
    weights = [1.0 / ((i + 1) ** s) for i in range(n)]
    total = sum(weights)
    target = rng.random() * total
    acc = 0.0
    for i, w in enumerate(weights):
        acc += w
        if acc >= target:
            return i
    return n - 1


def sample_bimodal(
    rng: random.Random, p_long: float, short_mean: float, long_mean: float, cv: float
) -> float:
    """Mixture of a short and a long mode.

    Real inference traffic is a mixture, not a single distribution: interactive
    chat plus long-document work in the same queue. The mixture is what creates
    head-of-line blocking, so a unimodal workload will hide the problem you are
    trying to study.
    """
    mean = long_mean if rng.random() < p_long else short_mean
    return lognormal(rng, mean, cv)

File: crates/sim-core/src/rng.rs
//! Named, independently seeded random streams.
//!
//! Why named rather than one global generator: if arrivals, request sizes and failure injection all
//! draw from one stream, then changing the arrival rate shifts every other draw too, and an A/B
//! comparison silently compares two different workloads. With named streams, enabling failure
//! injection cannot perturb the workload.
//!
//! Stream seeds come from a fixed hash of the name mixed with the master seed, so they are stable
//! across processes and platforms.

/// xoshiro256++ over a splitmix64-expanded seed. Small, fast, and deterministic everywhere.
#[derive(Clone)]
pub struct Rng {
    s: [u64; 4],
}

impl Rng {
    pub fn from_seed(seed: u64) -> Self {
        let mut z = seed;
        let mut s = [0u64; 4];
        for slot in s.iter_mut() {
            z = z.wrapping_add(0x9E37_79B9_7F4A_7C15);
            let mut x = z;
            x = (x ^ (x >> 30)).wrapping_mul(0xBF58_476D_1CE4_E5B9);
            x = (x ^ (x >> 27)).wrapping_mul(0x94D0_49BB_1331_11EB);
            *slot = x ^ (x >> 31);
        }
        Rng { s }
    }

    #[inline]
    pub fn next_u64(&mut self) -> u64 {
        let result = self.s[0]
            .wrapping_add(self.s[3])
            .rotate_left(23)
            .wrapping_add(self.s[0]);
        let t = self.s[1] << 17;
        self.s[2] ^= self.s[0];
        self.s[3] ^= self.s[1];
        self.s[1] ^= self.s[2];
        self.s[0] ^= self.s[3];
        self.s[2] ^= t;
        self.s[3] = self.s[3].rotate_left(45);
        result
    }

    /// Uniform in [0, 1).
    #[inline]
    pub fn f64(&mut self) -> f64 {
        // 53 significant bits, the most an f64 can hold exactly.
        (self.next_u64() >> 11) as f64 * (1.0 / (1u64 << 53) as f64)
    }

    #[inline]
    pub fn below(&mut self, n: u64) -> u64 {
        if n == 0 {
            return 0;
        }
        self.next_u64() % n
    }

    /// Exponential with the given mean. Interarrival time of a Poisson process.
    pub fn exponential(&mut self, mean: f64) -> f64 {
        if mean <= 0.0 {
            return 0.0;
        }
        // 1 - u avoids ln(0).
        -mean * (1.0 - self.f64()).ln()
    }

    fn normal(&mut self) -> f64 {
        // Box-Muller. One of the pair is discarded, which costs nothing that matters here and
        // avoids carrying state that would complicate reproducibility.
        let u1 = 1.0 - self.f64();
        let u2 = self.f64();
        (-2.0 * u1.ln()).sqrt() * (std::f64::consts::TAU * u2).cos()
    }

    /// Lognormal with the given arithmetic mean and coefficient of variation.
    ///
    /// Parameterised by mean and CV rather than by the underlying mu and sigma, because those are
    /// the numbers a workload is actually described by. Token counts are strongly right-skewed in
    /// practice, and `docs/calibration.md` measures output CVs above 3.
    pub fn lognormal(&mut self, mean: f64, cv: f64) -> f64 {
        if mean <= 0.0 {
            return 0.0;
        }
        if cv <= 0.0 {
            return mean;
        }
        let sigma2 = (1.0 + cv * cv).ln();
        let sigma = sigma2.sqrt();
        let mu = mean.ln() - sigma2 / 2.0;
        (mu + sigma * self.normal()).exp()
    }
}

/// One `Rng` per name, derived from a master seed.
pub struct Streams {
    master: u64,
}

impl Streams {
    pub fn new(master: u64) -> Self {
        Streams { master }
    }

    /// FNV-1a over the name, mixed with the master seed. Stable across processes, unlike a
    /// language-provided string hash, which would make runs irreproducible.
    pub fn stream(&self, name: &str) -> Rng {
        let mut h: u64 = 0xcbf2_9ce4_8422_2325;
        for b in name.as_bytes() {
            h ^= *b as u64;
            h = h.wrapping_mul(0x1000_0000_01b3);
        }
        Rng::from_seed(h ^ self.master.wrapping_mul(0x9E37_79B9_7F4A_7C15))
    }

    /// A per-entity stream, so replica 3's failures are independent of replica 4's.
    pub fn stream_indexed(&self, name: &str, index: u64) -> Rng {
        let mut r = self.stream(name);
        Rng::from_seed(r.next_u64() ^ index.wrapping_mul(0xD6E8_FEB8_6659_FD93))
    }
}
